File: data/readable_format/cpp/file1952.cpp
#include <iostream>
#include <algorithm>

// Assuming WFN1_AD_DBLE is essentially a double for simplicity.
// If it's more complex, define the structure or class accordingly.
using WFN1_AD_DBLE = double;

class WFN1_AD1 {
public:
    static int WFN1_AD_ILADLR(int M, int N, WFN1_AD_DBLE** A, int LDA) {
        const WFN1_AD_DBLE ZERO = 0.0;
        if (M == 0) {
            return M;
        } else if (A[M-1][0] != ZERO || A[M-1][N-1] != ZERO) { // Adjusting for 0-based indexing
            return M;
        } else {
            int result = 0;
            for (int j = 0; j < N; ++j) {
                int i = M;
                while (i >= 1 && A[std::max(i-1, 0)][j] == ZERO) { // Adjust for 0-based indexing
                    --i;
                }
                result = std::max(result, i);
            }
            return result;
        }
    }
};

int main() {
    // Example usage
    int M = 3;
    int N = 2;
    int LDA = 3;

    // Dynamic allocation of the 2D array A
    WFN1_AD_DBLE** A = new WFN1_AD_DBLE*[LDA];
    for (int i = 0; i < LDA; ++i) {
        A[i] = new WFN1_AD_DBLE[N];
    }

    // Initialize A with some values for testing
    for (int i = 0; i < LDA; ++i) {
        for (int j = 0; j < N; ++j) {
            A[i][j] = 0.0; // Fill with zeros or other values as needed
        }
    }
    // Assuming specific values that could trigger certain conditions
    A[M-1][0] = 0.0; // Remember, C++ uses 0-based indexing
    A[M-1][N-1] = 0.0;

    int result = WFN1_AD1::WFN1_AD_ILADLR(M, N, A, LDA);
    std::cout << "Result = " << result << std::endl;

    // Clean up memory
    for (int i = 0; i < LDA; ++i) {
        delete[] A[i];
    }
    delete[] A;

    return 0;
}

File: data/readable_format/fortran/file1952.f90
PROGRAM Main
  IMPLICIT NONE
  INTEGER :: M, N, LDA, result
  INTEGER, PARAMETER :: MAX_SIZE = 100
  DOUBLE PRECISION, ALLOCATABLE :: A(:,:)

  ! Example parameters - adjust as necessary
  M = 3
  N = 2
  LDA = 3
  ALLOCATE(A(LDA,N))

  ! Initialize the matrix A with some values for testing
  A = 0.0
  A(3,1) = 0.0
  A(3,2) = 0.0

  result = WFN1_AD_ILADLR(M, N, A, LDA)
  PRINT *, "Result = ", result

  DEALLOCATE(A)

CONTAINS

  INTEGER FUNCTION WFN1_AD_ILADLR(M, N, A, LDA)
    INTEGER, INTENT(IN) :: M, N, LDA
    DOUBLE PRECISION, INTENT(IN) :: A(LDA,*)
    INTEGER :: I, J
    DOUBLE PRECISION :: ZERO
    ZERO = 0.0D+0

    IF (M == 0) THEN
       WFN1_AD_ILADLR = M
       RETURN
    ELSE IF (A(M, 1) /= ZERO .OR. A(M, N) /= ZERO) THEN
       WFN1_AD_ILADLR = M
       RETURN
    ELSE
       WFN1_AD_ILADLR = 0
       DO J = 1, N
          I = M
          DO WHILE (A(MAX(I,1),J) == ZERO .AND. I >= 1)
             I = I - 1
          END DO
          WFN1_AD_ILADLR = MAX(WFN1_AD_ILADLR, I)
       END DO
    END IF
  END FUNCTION WFN1_AD_ILADLR

END PROGRAM Main